crates/aptos-rosetta/src/construction.rs

```diff
@@ -313,101 +288,160 @@ async fn construction_metadata(
         _ => {}
     }
 
-    // If both are present, we skip simulation
-    let (suggested_fee, gas_unit_price, max_gas_amount) =
-        if let (Some(gas_unit_price), Some(max_gas_amount)) = (
-            request.options.gas_price_per_unit,
-            request.options.max_gas_amount,
-        ) {
-            let suggested_fee = Amount::suggested_gas_fee(gas_unit_price.0, max_gas_amount.0);
-            (suggested_fee, gas_unit_price.0, max_gas_amount.0)
+    Ok(internal_operation)
+}
+
+async fn simulate_transaction(
+    rest_client: &aptos_rest_client::Client,
+    chain_id: ChainId,
+    options: &MetadataOptions,
+    internal_operation: &InternalOperation,
+    sequence_number: u64,
+) -> ApiResult<(Amount, u64, u64)> {
+    // If we have any missing fields, let's simulate!
+    let mut transaction_factory = TransactionFactory::new(chain_id);
+
+    // If we have a gas unit price, let's not estimate
+    let mut estimate_gas_price = true;
+    if let Some(gas_unit_price) = options.gas_price_per_unit.as_ref() {
+        estimate_gas_price = false;
+        transaction_factory = transaction_factory.with_gas_unit_price(gas_unit_price.0);
+    }
+
+    // Build up the transaction
+    let (txn_payload, sender) = internal_operation.payload()?;
+    let unsigned_transaction = transaction_factory
+        .payload(txn_payload)
+        .sender(sender)
+        .sequence_number(sequence_number)
+        .build();
+
+    // Read and fill in public key as necessary, this is required for simulation!
+    // TODO: Only single signer supported
+    let public_key =
+        if let Some(public_key) = options.public_keys.as_ref().and_then(|inner| inner.first()) {
+            Ed25519PublicKey::from_encoded_string(&public_key.hex_bytes).map_err(|err| {
+                ApiError::InvalidInput(Some(format!(
+                    "Public key provided is not parsable {:?}",
+                    err
+                )))
+            })?
         } else {
-            // If we have any missing fields, let's simulate!
-            let mut transaction_factory = TransactionFactory::new(server_context.chain_id);
+            return Err(ApiError::InvalidInput(Some(
+                "Must provide public_keys for simulation otherwise it can't simulate!".to_string(),
+            )));
+        };
 
-            // If there's a gas unit price we're using it, max gas doesn't matter the API will overwrite it
-            if let Some(gas_unit_price) = request.options.gas_price_per_unit {
-                transaction_factory = transaction_factory.with_gas_unit_price(gas_unit_price.0)
-            }
+    // Sign the transaction with a dummy signature of all zeros as required by the API
+    let signed_transaction = SignedTransaction::new(
+        unsigned_transaction,
+        public_key,
+        Ed25519Signature::try_from([0u8; 64].as_ref()).expect("Zero signature should always work"),
```

Comment:
Usually we have a method for this kind of thing, do we not have one here? Like `Ed25519Signature::zero`.